File: problems-adapt/0000-0100/0068_maximum-non-adjacent-loot/solution_dp.ts
function maxNonAdjacentLoot(nums: number[]): number {
    // Rolling two-variable DP: cur = best through position i-1, prev = best
    // through position i-2; both start at 0 ("nothing taken yet").
    let prev = 0,
        cur = 0;
    for (const x of nums) {
        // Skip this position (keep cur) or take it (prev + x).
        const next = Math.max(cur, prev + x);
        prev = cur;
        cur = next;
    }
    return cur;
}

File: problems/0100-0200/0198_house-robber/solution_dp.java
class Solution {

    public int rob(int[] nums) {
        // Rolling two-variable DP: cur = best through house i-1, prev = best
        // through house i-2; both start at 0 ("nothing robbed yet").
        int prev = 0,
            cur = 0;
        for (int x : nums) {
            // Skip this house (keep cur) or rob it (prev + x).
            int next = Math.max(cur, prev + x);
            prev = cur;
            cur = next;
        }
        return cur;
    }
}
